### ISSUE TITLE
Hololens2 Not responding to SDP offer and no Video Rendering

### ISSUE THREAD

**Issue by cherrysodalady** [microsoft/MixedReality-WebRTC#790]
**Describe the bug**
I have setup my unity in accordance to the tutorial located : https://microsoft.github.io/MixedReality-WebRTC/versions/release/2.0/manual/unity/helloworld-unity.html . I am able to get local video on my unity player, i am also able to see the exchange of GET datas 404 between the hololens and my laptop in the dss. 

i have created a button in my unity project which essentially just calls peerconnection.StartConnection() , when i call this the SDP offer message can be seen but then the hololens drops out of the dss and nothing more happens. 

The problem is i am unable to see any video on my hololens2. 

**SDP messages**
first picture is before the button is pressed and the second two are after the button is pressed. The button is pressed on the laptop and contains the code in the following screenshots

Dss before button click
![image](https://user-images.githubusercontent.com/41712449/125888802-29378a79-23eb-4af3-a970-cc2291e47846.png)

button click code
![image](https://user-images.githubusercontent.com/41712449/125888862-4965e8e0-9f3d-45ad-9c58-49810e5d80f6.png)

Dss After Button click
![image](https://user-images.githubusercontent.com/41712449/125888881-8cf4f675-7341-4182-a640-a043a56869e2.png)

![image](https://user-images.githubusercontent.com/41712449/125888938-25c35e90-5df6-482a-8e4e-be18c13a057e.png)

**Screenshots**
here are screenshots of what has been configured in unity 

![localcapture](https://user-images.githubusercontent.com/41712449/125889012-8050997d-42ff-42d7-a246-4d1a5cb97f7d.PNG)

![peerconnection](https://user-images.githubusercontent.com/41712449/125888978-99161d12-f2df-4de7-9fcf-4731d72d4fcc.PNG)

![signaller](https://user-images.githubusercontent.com/41712449/125889000-df0e1ce2-45b5-4777-b097-4be1f98498de.PNG)

![localvideoplayer](https://user-images.githubusercontent.com/41712449/125889055-d1d1a9a8-8707-488b-be02-5e740aa17b1d.PNG)

![remotevideo](https://user-images.githubusercontent.com/41712449/125889064-1a33137f-611c-4c2d-a5c6-451fe4d86c26.PNG)

![micsource](https://user-images.githubusercontent.com/41712449/125889070-7103598a-003a-4671-96d1-5ca5c69a5b9f.PNG)

![remoteaudio](https://user-images.githubusercontent.com/41712449/125889077-9f5b9941-14dd-462c-b46e-62628bd279ab.PNG)

**Environment**
_Please fill the information for **each peer** if different_
 - MR-WebRTC version 2.0.2
 - Platform: [e.g. Windows Desktop, Unity Editor, and Hololens 2
 - Unity version (if applicable): 2020.3.11.f1
 - Target device: Hololens2

**Comment by spacecheeserocks** [comment]
This could be a firewall issue with your PC - try (temporarily) disabling your Windows Firewall and any third party firewall software.

If this works, you may want to investigate running a TURN server, which provides an alternative path when peers are unable to directly communicate.

Another thing to check: if you have enforced H264 video codec for HoloLens performance, then try temporarily disabling that. The unity editor/windows desktop versions **cannot** handle H264 video (they only support VP8).

**Comment by cherrysodalady** [comment]
i have indeed disabled the firewalls on my PC (temporarily) and notice the same result, in terms of the video codec, i just went and had a look within the webcam source script and found the script i feel may be responsible (which is commented out) would you be able to provide some guidance as to what is supposed to be here? if its supposed to change. Thank you 
![image](https://user-images.githubusercontent.com/41712449/126084104-1ae525e8-47f8-40ef-9d17-b40400618732.png)

**Comment by Matteo-0** [comment]
Hi @cherrysodalady, where you able to solve this problem? I have actually the same problem as you.
I am trying to send local video from unity to the hololens.
If you have any advice please let me know.

Thank you

**Comment by Matteo-0** [comment]
How do I change Videocodec? I am looking everywhere but cannot find anything.
I also decommented those lines of code but I get a lot of errors.

**Comment by SilverLive** [comment]
Maybe the Problem is that the VideoFrameCodec is different from the one which is attached to the VideoRenderers. If your video source delivers VideoFrames with codec I420, the renderer which displays the video has to use YUVFeedMaterial. If it delivers ARGB32 you need to use the ARGBFeedMaterial to render. Unfortunately i didn´t find any possibility to make ARGB32-based  VideoSources work in MR WebRTC. It seems like the native peers always use I420 codec and I didn´t find any way to change that behaviour. But since you are using WebcamSource it should already be I420 so your Renderer should all use YUVFeedMaterial. Maybe this is your problem. Tell me if it was :)

**Comment by Matteo-0** [comment]
Thanks for the answer. Yes I am using WebcamSource and my Renderer uses YUVFeedMaterial but nothing works.
The problem appears when I press the button "CreatOffer" in unity to start the connection. When I press that button my hololens disconnect from node dss and I do not know why. Also another problem is that on the local screen on the hololens I should see what my hololens camera is seeing but instead I get a black screen. 
Do not know how to proceed I really tried everything.

**Comment by SilverLive** [comment]
I tried it myself and unfortunately have the same problem. No Webcam-Stream at all.. only black screens.. So far I also have no idea about the reason for that.

**Comment by cherrysodalady** [comment]
hi @Matteo-0  - i have not solved the problem yet, i have also not used "CreateOffer" but instead "StartConnection()" as can be seen in my picture, is this different to what you  have? and yes i dont see any video from my hololens either its black like yours 

Hi @SilverLive i am still unsure how to change the video codecs, would you be able to explain how to do this? or @spacecheeserocks also suggested this but im not sure how to go about doing it. Thank you :)

**Comment by Matteo-0** [comment]
Could it be that I am building my application for hololens 2 with ARM64 and recently I read that WebRTC is not compatible with ARM64? Today I will try building it with ARM, hope it will work, if it works I will lwt you know.

**Comment by SilverLive** [comment]
@cherrysodalady you have to call StartConnection from the peer which should set up the offer. You can only call it after both client peers were Initialized. StartConnection triggers the whole pairing/signaling process. So here you are right. 

Unfortunately also didn´t find any code which allows changing the videoEncoding. I only mentioned VideoFrameCodecs and in this case you are limited to I420 or ARGB32. 

@Matteo-0 You are totally right. I tested it and with ARM it works! Thank you, I already read that a view days ago but i didn´t have that in mind ;)

**Comment by spacecheeserocks** [comment]
> Could it be that I am building my application for HoloLens 2 with ARM64 and recently I read that WebRTC is not compatible with ARM64?

@Matteo-0 Correct, unfortunately there is no ARM64 version of this library currently - you can track any progress at #414. As this repo is currently unmaintained (#773), I expect that an ARM64 version won't be appearing any time soon, unless someone makes a PR for it of course :)

**Comment by Matteo-0** [comment]
Ok, It worked for me too, I am so happy and happy for you too @SilverLive! 
Can anyone please explain me what is the difference between building an application with ARM or ARM64? @SilverLive @spacecheeserocks @cherrysodalady

**Comment by cherrysodalady** [comment]
hi @Matteo-0 and @SilverLive  i changed to ARM and theres definately progress! i now see the hololens's video on the hololens, i still however dont see the webcam feed from my laptop on the hololens (so i can see only one video not two) are you guys able to see the two?

**Comment by spacecheeserocks** [comment]
> Can anyone please explain me what is the difference between building an application with ARM or ARM64?
In this case, it's non trivial. 

You need to rebuild the underlying native library (libwebrtc), which requires **a lot** of patience (and disk space!). It requires a slightly obscure dev environment, and some customised C++ build tools.

Additionally, the code as-is just fails to build for ARM64. I'm not 100% sure why, but I think some of the dependencies fail the build process because there are no ARM64 versions available. It may also be that there is hardware-specific ARM-specific code blocks, which would need to be rewritten for ARM64. There has already been some discussion about this in other issues, but I don't remember where exactly (probably #414).

The docs do give *some* instructions for building the native library from scratch, and there are some other instructions to follow for building webrtc-uwp underneath.

TL;DR, if you are comfortable with low level C++, as well as writing assembly code for ARM/ARM64, then you could probably give it a try.

**Comment by Matteo-0** [comment]
> hi @Matteo-0 and @SilverLive i changed to ARM and theres definately progress! i now see the hololens's video on the hololens, i still however dont see the webcam feed from my laptop on the hololens (so i can see only one video not two) are you guys able to see the two?
> 
> it may be something wrong with my remote video component (as it differs from the tutorial as the components have changed) is this what you have put or should it be different?
> 
> ![image](https://user-images.githubusercontent.com/41712449/126717667-64ba3eac-0587-418f-aaec-1de5a1c9856f.png)

I put in Video Stream Started: VideoRenderer -> StartRendering()

While in Video Stream Stopped: VideoRenderer -> StopRendering()

Try with this, hope it will work!

**Comment by cherrysodalady** [comment]
Hi so does that mean U have the webcam source in the other box too? Instead
of Remotevideoplayer?

**Comment by cherrysodalady** [comment]
omg its all working! thank you so much everyone for your help, really appreciate it :)
